js/classes/entities/foreground/precipitation.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * Creates falling rain or snow.
+ * 
+ * @author Caleb Krauter
+ */
+class Precipitation {
+
+    /**
+     * Based on the parameters a single raindrop or snowflake is added creating precipitation.
+     * 
+     * @param {*} dir direction that the drop should be falling.
+     * @param {*} pos initial position of the drop.
+     * @param {*} type rain or snow.
+     * @param {*} scale scale of the rain or snow will vary.
+     */
+    constructor(dir, pos, type, scale) {
+        this.dir = dir;
+        this.origin = pos;
+        // Setting this.pos to (0, 0) makes rain/snow fall naturally when it starts falling.
+        this.pos = new Vector(0, 0);
+        this.type = type;
+        this.velocity = new Vector(0, 0);
+        this.scale = scale;
+        this.imgIndex = 0;
+        // Precipitation.SPRITESHEET = Precipitation.SPRITESHEET_RAIN_DOWN;
+
+        if (type === "rain") {
+            if (dir === "down") {
+                this.imgIndex = 0;
+            } else if (dir === "left") {
+                this.imgIndex = 1;
+            } else if (dir === "right") {
+                this.imgIndex = 2;
+            }
+        } else if (type === "snow") {
+            let varient = Math.ceil(Math.random() * 4);
+            switch (varient) {
+                case 1:
+                    this.imgIndex = 3;
+                    break;
+                case 2:
+                    this.imgIndex = 4;
+                    break;
+                case 3:
+                    this.imgIndex = 5;
+                    break;
+                case 4:
+                    this.imgIndex = 6;
+                    break;
+                default:
+                    break;
+            }
+
+        }
+        this.animator = new Animator(Precipitation.SPRITESHEET, new Vector(0, this.imgIndex * Precipitation.SIZE.x), new Vector(Precipitation.SIZE.x, Precipitation.SIZE.y), 1, 1);
+        this.hasStartedRaining = false;
+        this.stopRain = false;
+    }
+
+    /**
+     * Image used for precipitation.
+     */
+    static get SPRITESHEET() {
+        return "./sprites/snow_and_rain.png";
+    }
+
+    /**
+     * Size used for the image drawn.
+     */
+    static get SIZE() {
+        return new Vector(32, 32);
+    }
+
+    /**
+     * Resets the rain or snow to the top of the screen.
+     */
+    reset() {
+        let variationY = Math.random() * (50 + 100) - 50;
+        // Ensure that the rain/snow covers the screen when resetting its origin.
+        let variationX = Math.random() * (1920 + 3000);
+        // We use CHAD.pos.x - 1920 to ensure that the rain/snow follows Chad and that the rain/snow begins at the left of the screen as it covers it.
+        if (this.dir == "right") {
+            variationX = Math.random() * (1920 + 3000);
```

Comment:
These numbers (1920, 3000, etc) are kind of mysterious, and it may be good to use constants to make it clear what they represent.

Reply:
Good point, I'll update that!